src/self-update/README.md

```diff
@@ -0,0 +1,89 @@
+
+* A softlink to be created in `/etc/tedge/sm-plugins/` named `tedge` that points to `apt` plugin, will make the SM agent aware.
+
+* The plugin managing module type "tedge" must take all packages of an update request at once from SM Agent (i.E. Plugin API command "update-list" must be used). That is to get the plugin independent from potential restarts of the SM agent (those restarts are part of updating the SM agent).
+
+* Before any package update for module type "tedge" starts, all needed installation resources for the entire update request must be downloaded (i.E. all \*.deb packages for APT). That is to get the plugin uncoupled from any potential network issue during update procedure.
+  * For all packages with some URL given in the update request the SM Agent manages download before plugin start, and provides local file paths to the plugin.<br/>
+    TODO: What happens if one package download fails in SM agent? Does plugin maybe need to check if all files are locally available?
+  * For all packages without any URL given in the update request the Package Manager (i.E. APT) will manage the download as part of the update. Therefor the plugin must instruct the package manager to first download **all** packages. 
+    * If all packages were successfully downloaded, the package manager can start updating packages.<br/>
+    * Instead, if one or more download fails the plugin must stop before any installed package was touched by update, and report the entire update process as failed to the SM agent<br/>
+
+* After processing the complete update request the plugin will publish the overall exit code as MQTT retain message to topic `tedge/plugins/software/<plugin name>`, and exits.
+
+* As both module types "tedge" and "apt" use same package manager APT, the plugin manages a blacklist to allow later to distinguish packages reported by Package Managers `apt list` command.
+  * The blacklist is a TOML file that stores package names of all packages that were installed/updated with module type "tedge".
+  * When the plugin manages an update request for a package with module type "tedge" it adds the package to the blacklist.
+  * When the plugin manages an update request for a package with module type "apt" it removes the package from the blacklist.
+  * When the plugin is called with LIST command for module type "apt", all non tedge packages (all installed but not blacklisted ones) will be reported.
+  * When the plugin is called with LIST command for module type "tedge", all tedge packages (all installed and blacklisted ones) will be reported.
+
+
+### Details about the SM Agent
+
+* The SM agent must accept the final exit code of a plugin execution via MQTT retain message on topic `tedge/plugins/software/<plugin name>`, send according update result message to the mapper, and clear the retain message on the topic afterwards.
```

Comment:
To make things more understandable, I would :
* tell why we are doing this
* group here the responsibilities related both to the plugin and the agent

Reply:
Is there an intention to preserve and extend current exit codes or just change for all plugins?

Reply:
I added more details about that aspect to spec, as below.

Please check if that answers your points:


* Transfer of plugins exit code to the SM Agent:<br/>
  NOTE: During execution of plugin for module type „tedge“ it is valid (or even expected) that the SM Agent will restart. The restarted SM Agent process will be disconnected from still running plugin's process and it's exit code.

  * For module type tedge the SM agent must accept the final exit code of a plugin execution via MQTT retain message on topic `tedge/plugins/software/tedge`, send according update result message to the mapper, and clear the retain message on the topic afterwards.
  * It might happen that the SM Agent does not restart and keeps connected to the plugin for module type "tedge" (e.g. when agent is not part of the update request, or the update fails). Also then the exit code that is reported by the plugin via MQTT must be used, instead of the exit code reported from plugin's process exit.